### ISSUE TITLE
Patching for a large slice of state is slow

### ISSUE THREAD

**Issue by ethanroday** [tshaddix/webext-redux#153]
Consider a fairly standard key-based state like the following:

    items: {
      "a": { ... },
      "b": { ... },
      "c": { ... },
      "d": { ... },
      // ...
    }

My reducer for updating an item will look something like the this:

    const itemsReducer = (state = {}, action) => {
      switch(action.type) {
        case UPDATE_ITEM:
          return {
            ...state,
            [action.id]: action.item
          }
        // ...
      }
    }

    const reducer = combineReducers({
      items: itemsReducer,
      // ...
    })

The problem is that `shallowDiff` picks up on the fact that `state.items` is now a new object, so it patches the _entire_ `state.items` object, even though it was really only one key that was updated. If `state.items` is large, then this can be a significant performance issue.
How can this be handled? Basically, `patchState` (or, more specifically, `shallowDiff`) needs to "know" which keys are just defining state shape and which keys are actual data. I suppose this is not impossible - it's basically [what `combineReducers` does](https://github.com/reduxjs/redux/blob/master/src/combineReducers.js), but it would require more configuration from the user.
Another more radical idea is to push all state back into the background page of the extension, and then force the popup and content scripts to access state using selectors. The selectors could then be memoized in the popup and content scripts, and be proactively updated from the background only when the relevant state changed. This would obviously have its cons, chief among which would be the fact that applying middleware from the proxy store would no longer really make very much sense.

**Comment by ethanroday** [comment]
Thinking a little more about this, I guess it comes down to a tradeoff between efficiency of _computing_ patches and efficiency of _communicating_ patches. The current method of computing patches - `shallowDiff` - is fast, but it also overcommunicates. More complex patch computation logic might make for smaller patches, but they would take longer to compute. The goal in the scenario above is to only send a patch that looks something like this (assuming item `d` is updated):

    {
      items: [
        key: "d",
        value: { ... },
        change: "removed"
      ]
    }

This at least requires iterating through all of the keys of `state.items`.

**Comment by ethanroday** [comment]
Started working on a [branch](https://github.com/ethanroday/react-chrome-redux/tree/deep-diffs) for this. It's untested and largely undocumented, but it's a start. The basic idea is that a user can specify different diff strategies, which will be used in the background to generate state patches and in the content scripts to patch the state. There are two built-in strategies: shallow diff (which is the default) and deep diff, which will keep diffing down the tree until it gets to things that are not plain objects.
This can also potentially solve #144. See my comment there.
@tshaddix or @vhmth, thoughts?

**Comment by vhmth** [comment]
@ethanroday the commits on your local branch are **really** exciting. I'm wondering how we missed this with the shallow diff approach. I love how you made it so you can choose a diff strategy as well. If we can get some tests around the deep diff strategy, I'm cool with defaulting to shallow diff (to be conservative about edge cases that might pop up with deep diffing) and allow people to choose a diff scheme with the plug-and-play nature of what you're coding up.

Thoughts @tshaddix?

**Comment by ethanroday** [comment]
@tshaddix, I've created a PR for this one: #154. Let me know what you think!

**Comment by marcokam** [comment]
@ethanroday I'm excited by your PR, thanks for the great contribution!

I'm curious about the strategy you are using to utilize deep diffing in your own code.  What do you set the `shouldContinue` function to?

In the example you share above with the `items` object, I understand that if a single key in the item changed, the deep diff strategy would be great in that only the specific key would be patched.  What happens on initialization of items?  I can imagine that there may be a sever call returning an array of 100 items, the client side would then normalize and update the store.  In that case, wouldn't deep diff result in 100 patches for each key in item?  Or do you have some strategy using shouldContinue to handle this case so that on initialization the entire items object gets a single patch?

**Comment by ethanroday** [comment]
Thanks @marcokam! To your questions:

Basically, I figured `shouldContinue` would be useful for cases where you know it's not worth it to check individual properties on an object. It's not the most elegant solution, and I'm definitely open to suggestions for a better design. (One thing that I've noticed as I've used it is that it might be better to reverse the condition (i.e. make it `shouldStop` instead of `shouldContinue`), since it seems a little easier to think about it that way.)
For a practical example, let's say you were doing some network monitoring and you had a store that looked something like this:

```
{
  machines: {
    "my-machine-name": {
      ip: "0.0.0.0",
      uptime: 1824, // mins
      numActiveConnections: 55,
      avgUploadSpeed: 321.6, // Mbps
      avgDownloadSpeed: 513.2, // Mbps
      avgLatency: 12, // ms
      // ...
    },
    // ...
  },
  // ...
}
```

When you update the info for a given machine, it's likely that most of the properties are changing slightly every time. So rather than check each one, you can set `shouldContinue` to `(oldState, newState, context) => !(context.length === 2 && context[0] === "machines")`, and the deep diff algorithm won't check the individual properties for each machine.
I'm using it in my code for a similar situation, except it's statistics about pages the user has visited rather than network connections.

----------------

As long all 100 items are coming in in a single action, then there will still only be a single patch message, but you're right that it would be more verbose than a shallow diff patch. For example, let's say there's an `ITEM_FETCH_SUCCESS` action that comes back with `a`, `b`, and `c` items. With shallow diff, the patch message would look like this:

```
[
  {
    "key": "items",
    "value": {
      "a": { ... },
      "b": { ... },
      "c": { ... }
    },
    "change": "updated"
  }
]
```

Whereas the deep diff patch would look like this:

```
[
  {
    "key": "items",
    "change": "updated_keys",
    "value": [
      {
        "key": "a",
        "change": "updated",
        "value": { ... }
      },
      {
        "key": "b",
        "change": "updated",
        "value": { ... }
      },
      {
        "key": "c",
        "change": "updated",
        "value": { ... }
      }
    ]
  }
]
```

The deep diff algorithm always treats new keys as updated values, and doesn't traverse down into their individual properties (because it knows there's nothing to compare them to in the old state).

**Comment by marcokam** [comment]
@ethanroday thanks for your explanation, it makes more sense to me now.  I didn't really have an issue with thinking about it either as `shouldContinue` or `shouldStop`, but I was trying to think if there would be a more generic solution to figuring out if the deep diff should continue or not.  It sounds like you may be using it for a specific case (when you know which exact keys you may want to continue/stop at), which is fine, and at the moment I can't think of a more generic solution either.

I think the current implementation may produce a diff with updates to send over to the proxy store that may not be necessary in certain cases.  Assuming we just have `shouldContinue` set as the default (always recursively diff objects), and using the same `items` object above, let's say we make a call to the server to get new items.  After the server call is done, we normalize and update the store.  If the server results aren't actually different from what's currently in the store, deepDiff would give something like this:

```
[
  {
    "key": "items",
    "change": "updated_keys",
    "value": [
      {
        "key": "a",
        "change": "updated_keys",
        "value": []
      },
      {
        "key": "b",
        "change": "updated_keys",
        "value": []
      },
      {
        "key": "c",
        "change": "updated_keys",
        "value": []
      }
    ]
  }
]
```

If we only include diffs for "updated_keys" when value.length !== 0, couldn't we completely ignore any patching for `items` in this case?

In practice, it may not matter too much, since the patching strategy checks for `difference.length`, but as you mentioned in your original comment, in this case we may be _overcommunicating_ the patch.

**Comment by ethanroday** [comment]
As far as genericness is concerned I completely agree, but it's really very context-dependent when the diffing should stop, so I went with something that, while a little unwieldy, doesn't place any restrictions on what logic the user can implement. One thought I had originally to make it easier to use would be to allow the user to specify a sort of partial state tree indicating the paths where the diff should continue. Each diff call would recurse down the tree and stop when it got to a leaf. But I didn't get very far in the implementation, and I decided it would actually be pretty cumbersome for situations where you only wanted to stop in a couple of cases and redundant in situations where there were multiple cases with parallel logic but in different parts of the tree.

Actually, the behavior you point out with the empty values is intentional. The deep diff algorithm not only ensures that the states look the same, but also that object references are changed appropriately. In the situation you describe, the object references to `state.items`, `state.items.a`, `state.items.b`, and `state.items.c` have all changed in the background, so they need to change in the proxy stores as well.

**Comment by marcokam** [comment]
I had the same thought about keeping the object references in sync after I sent my earlier comment.  It makes sense, thanks!

**Comment by tshaddix** [comment]
@ethanroday Just getting back from a month of traveling for work now and diving into the PR. Thank you so much for a) identifying the issue and b) proposing multiple solutions. Going through strategies now.

**Comment by tshaddix** [comment]
This has been published under `2.0.0-alpha.2`. Thanks for your excellent work!

**Comment by tshaddix** [comment]
Now available in `webext-redux@2.0.0` :)
